lib/services/local/provisioning_states.go

```diff
@@ -0,0 +1,133 @@
+// Teleport
+// Copyright (C) 2024 Gravitational, Inc.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package local
+
+import (
+	"context"
+
+	provisioningv1 "github.com/gravitational/teleport/api/gen/proto/go/teleport/provisioning/v1"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/backend"
+	"github.com/gravitational/teleport/lib/services"
+	"github.com/gravitational/teleport/lib/services/local/generic"
+	"github.com/gravitational/teleport/lib/utils/pagination"
+	"github.com/gravitational/trace"
+)
+
+const (
+	provisioningStatePrefix   = "provisioning_state"
+	provisioningStatePageSize = 100
+)
+
+type ProvisioningStateServiceMode int
+
+const (
+	// ProvisioningStateServiceModeStrict is the default service mode, with
+	// strict validation enabled.
+	ProvisioningStateServiceModeStrict ProvisioningStateServiceMode = 0
+
+	// ProvisioningStateServiceModeRelaxed indicates that the service should do
+	// no validation and just write to the provided backend. This is generally
+	// for use with caches
+	ProvisioningStateServiceModeRelaxed ProvisioningStateServiceMode = 1
```

Comment:
This is a little strange. Can you explain what this is for? I'm not sure why we would be validating things when writing downstream but not cache. Is this a pattern we already use elsewhere?

Reply:
I wanted to use the same service for primary access and caching, but the cache instances failed with `ComparisonErrors` when I used `ConditionalUpdateResource()`. This switches it to an unconditional update when used in the cache.

I'm definitely open to better suggestions, as this has always felt like a cop-out. How do other services that rely on optimistic locking work with the cache?

Reply:
@rosstimothy Any suggestions here?

Reply:
Only that the cache kept crashing due to Comparison errors when I was watching for changes on Provisioning State records.

To be fair:

1. I didn't spend a heap of time looking into the underlying case while building the demo this came from; it seemed plausible that a cache update might occasionally skip over a revision of a resource if the resource was being updated several times in quick succession, and getting a working demo was more important than 100% cache consistency at the time, also
2. I'm not super-familiar with our caching + event system, so I may just have been doing it wrong.

I know that the AccessList service does a similar thing (but for different reasons), but breaks it out into two separate service types - one service in `services/local` for auth to use, and another in `services/simple` for the cache.

Reply:
@tcsc Can you remove this relaxed/strict mode distinction and show the errors you were running into? Then we'll be able to better understand what the issue is.

Reply:
Will try to replicate and let you know.

Reply:
Example error. I read it as the cache write failing due to the `CompareFailedError` and then everything goes to pieces as the watchers fail and the cache resets.

```
2024-10-10T13:18:43+11:00 WARN [AUTH.ENTE] Watcher closed: event fanout system reset (retry in 5s) auth/usermonitor.go:211
2024-10-10T13:18:43+11:00 WARN  cert authority watch loop for client TLS config generator failed error:[
ERROR REPORT:
Original Error: *errors.errorString ca watcher exited with: event fanout system reset
Stack Trace:
        github.com/gravitational/teleport/lib/auth/client_tls_config_generator.go:248 github.com/gravitational/teleport/lib/auth.(*ClientTLSConfigGenerator).watchForCAChanges
        github.com/gravitational/teleport/lib/auth/client_tls_config_generator.go:196 github.com/gravitational/teleport/lib/auth.(*ClientTLSConfigGenerator).refreshClientTLSConfigs
        runtime/asm_arm64.s:1223 runtime.goexit
User Message: ca watcher exited with: event fanout system reset] auth/client_tls_config_generator.go:202
2024-10-10T13:18:43+11:00 ERRO [AUTH.ENTE] Error running access list ineligible reconciler error:[
ERROR REPORT:
Original Error: *errors.errorString event fanout system reset
Stack Trace:
        github.com/gravitational/teleport/e/lib/accesslist/reconciler_ineligible_status.go:160 github.com/gravitational/teleport/e/lib/accesslist.(*IneligibleStatusReconciler).Run
        github.com/gravitational/teleport/e/lib/accesslist/service.go:1967 github.com/gravitational/teleport/e/lib/accesslist.(*Service).runAccessListIneligibleReconciler.func1
        github.com/gravitational/teleport/lib/backend/helpers.go:228 github.com/gravitational/teleport/lib/backend.RunWhileLocked
        github.com/gravitational/teleport/e/lib/accesslist/service.go:1941 github.com/gravitational/teleport/e/lib/accesslist.(*Service).runAccessListIneligibleReconciler
        runtime/asm_arm64.s:1223 runtime.goexit
User Message: event fanout system reset] accesslist/service.go:1968
2024-10-10T13:18:43+11:00 WARN [IAM-IC:RM] Watcher has signaled exit pid:83125.1 identitycenter/resourcemonitor.go:135
2024-10-10T13:18:43+11:00 WARN [PROC:1]    Sync rotation state cycle failed. pid:83125.1 retry_interval:10s service/connect.go:659
2024-10-10T13:18:43+11:00 WARN [AUTH:1:CA] "Re-init the cache on error: updating provisioning state record\n\tresource revision does not match, it may have been concurrently created|modified|deleted; please work from the latest state, or use --force to overwrite" error:[
ERROR REPORT:
Original Error: *trace.CompareFailedError resource revision does not match, it may have been concurrently created|modified|deleted; please work from the latest state, or use --force to overwrite
Stack Trace:
        github.com/gravitational/teleport/lib/backend/memory/memory.go:419 github.com/gravitational/teleport/lib/backend/memory.(*Memory).ConditionalUpdate
        github.com/gravitational/teleport/lib/backend/report.go:255 github.com/gravitational/teleport/lib/backend.(*Reporter).ConditionalUpdate
        github.com/gravitational/teleport/lib/services/local/generic/generic.go:375 github.com/gravitational/teleport/lib/services/local/generic.(*Service[...]).ConditionalUpdateResource
        github.com/gravitational/teleport/lib/services/local/generic/generic_wrapper.go:138 github.com/gravitational/teleport/lib/services/local/generic.ServiceWrapper[...].ConditionalUpdateResource
        github.com/gravitational/teleport/lib/services/local/provisioning_states.go:91 github.com/gravitational/teleport/lib/services/local.(*ProvisioningStateService).UpdateProvisioningState
        github.com/gravitational/teleport/lib/cache/provisioning.go:69 github.com/gravitational/teleport/lib/cache.provisioningStateExecutor.upsert
        github.com/gravitational/teleport/lib/cache/genericcollection.go:89 github.com/gravitational/teleport/lib/cache.(*genericCollection[...]).processEvent
        github.com/gravitational/teleport/lib/cache/cache.go:1704 github.com/gravitational/teleport/lib/cache.(*Cache).processEvent
        github.com/gravitational/teleport/lib/cache/cache.go:1448 github.com/gravitational/teleport/lib/cache.(*Cache).fetchAndWatch
        github.com/gravitational/teleport/lib/cache/cache.go:1187 github.com/gravitational/teleport/lib/cache.(*Cache).update
        runtime/asm_arm64.s:1223 runtime.goexit
User Message: updating provisioning state record
        resource revision does not match, it may have been concurrently created|modified|deleted; please work from the latest state, or use --force to overwrite] cache/cache.go:1193
```

This is off my dev branch. You can find the cache setup code [here](https://github.com/gravitational/teleport/tree/86e4c6e417d62c408145c39d601194480da0921e/lib/cache).

Specific points of interest:
* Cache Executor: [lib/cache/provisioning.go](https://github.com/gravitational/teleport/blob/86e4c6e417d62c408145c39d601194480da0921e/lib/cache/provisioning.go)
* Collection Setup: [lib/cache/collections.go:675](https://github.com/gravitational/teleport/blob/86e4c6e417d62c408145c39d601194480da0921e/lib/cache/collections.go#L675)

Reply:
Thanks, @rosstimothy - that clarifies a lot.